### DISCUSSION TITLE
libp2p-rPGF-1: Nominations for Software & Dependencies

### DISCUSSION THREAD

**Discussion by p-shahi** [libp2p/libp2p#210] [RetroPGF]
### Software & Dependencies Nominations

This discussion post is for projects nominations to the Software & Dependencies category.
This category is intended for software that’s used to build libp2p and its protocols, projects that libp2p implementations/stacks depend on, and code contribution to protocols or standards that power libp2p.

### Nomination Template
When nominating a project, please answer the following questions in a comment to this discussion:

```
Project name:
The project name

Description:
A description of the positive impact of this project for this specific category over the past year (March 2023 - March 2024).

Github/Twitter link:
A link to the project’s GitHub or Twitter

Contact:
Contact info for the project or project lead
```

Things to consider when nominating a project:

- Describe how this project has delivered impact to the libp2p ecosystem in the previous year (March 2023 - March 2024). Use the provided template.
- Nominate projects in **one** relevant category. Do not nominate the same project across different categories! If a project has overlap in multiple categories, please choose just one category that best aligns with it. 
- Organizers of libp2p-RPGF-1 may reassign a project’s category from its nominated one if it better aligns elsewhere.
- Check if a project has been nominated already to avoid duplication.
- Please provide feedback to help the process improve in each round. Your input and feedback is valuable and you can use this post to submit feedback now.

Nominations will close on May 3rd, 2024.

See https://github.com/libp2p/libp2p/discussions/209 for an overview of the process, timelines, and to ask general questions.

**Comment by backkem** [comment]
Project name:
Pion WebRTC

Description:
This story starts with a [vision form 10 years ago](https://github.com/jbenet/random-ideas/issues/13). This helped inspire a community to start reading dozens of intertwined Web specs, RFCs and the Chromium implementation of WebRTC and slowly build out the entire stack in pure Go. Apart from fostering its own vibrant community, it [emboldened](https://github.com/libp2p/rust-libp2p/issues/1066) others to take on a [Rust port](https://github.com/webrtc-rs/webrtc). The initial [Rust transport landed](https://github.com/libp2p/rust-libp2p/pull/2622) at the end of 2022. Finally, the experimental [go transport landed](https://github.com/libp2p/go-libp2p/pull/2337) last year.

This is a story of persistence. The dedication shown by the Pion community mirrors that of LibP2P itself!

Github/Twitter link:
https://github.com/pion/webrtc
https://twitter.com/_pion

Contact:
@Sean-Der

**Comment by masih** [comment]
Project name:
[IPNI](https://github.com/ipni) and [HTTP over libp2p sync protocol](https://github.com/ipni/specs/blob/main/IPNI_HTTP_PROVIDER.md)

Description:
IPNI is a pioneering force in enhancing content discovery and retrieval across the Filecoin and IPFS networks. Through its innovative routing system, IPNI efficiently handles over 7 billion hashes daily, simplifying the process of accessing content through hash-based queries. To achieve scalability worldwide, IPNI relies heavily on libp2p as a dependable transport protocol, facilitating the synchronization of content hashes from numerous providers globally.

A significant milestone in IPNI's journey is its early embrace and integration of HTTP over libp2p. In spearheading this integration, IPNI has played a pivotal role in refining the protocol's specifications to seamlessly align with its sync mechanism. By customizing the HTTP over libp2p framework, IPNI has ensured the reliable ingestion of content from diverse providers, amplifying the efficiency of IPLD dag syncing. This collaborative endeavor not only fortifies the network's syncing capabilities but also sets new standards for scalability and adaptability.

Github/Twitter link:
* [IPNI](https://github.com/ipni) 
* [IPNI HTTP Provider Specification](https://github.com/ipni/specs/blob/main/IPNI_HTTP_PROVIDER.md) 
* [IPNI HTTP Sync Protocol Implementation](https://github.com/ipni/go-libipni/tree/main/dagsync/ipnisync)

Contact:
@masih @willscott @gammazero 

hi@ipni.io

**Comment by ianopolous** [comment]
Project name:
Peergos

Description:
We have implemented yamux, TLS, autonat, bitswap, http-over-libp2p and kademlia in Java/Kotlin.
These are all available in Nabu, and used in production in Peergos. Yamux, TLS and autonat are upstreamed into jvm-libp2p and have been deployed in Teku which runs ~15% of Ethereum nodes. During our work it became apparent that go-libp2p was dropping mplex support which would have cut off the jvm-libp2p implementations from the rest of the Ethereum network, until we implemented yamux. 

Github/Twitter link:
https://github.com/libp2p/jvm-libp2p
https://github.com/peergos/nabu
https://github.com/peergos/peergos
https://twitter.com/peergos

Contact:
ian@peergos.org

**Comment by Khriz01** [comment]
Project name:
Security Labs

Description:
A conceptual design that leverages libp2p to build a quantum-proof DePIN for sensitive data storage (e.g., private keys) through fully homomorphic-based mixnets and ZK-proofs. 

This system stores data anonymously and untraceably as a fully permissionless, trustless and decentralized authentication layer.  Authentication is described in their litepaper: https://outstanding-cellar-c8d.notion.site/Litepaper-6cc283133c974c0781aac956b0e6be09 being the short  short and intuitive description: “A decentralized cold-storage.”

Libp2p is a crucial component to ensure the transport layer is compatible and kept modular to interact with every other core system component: Mixing Layer, Application Layer, and Storage Layer.

The core technology and authentication framework can be found in the whitepaper: https://docsend.com/view/dbk48wukd3ivd3ad

Twitter: https://twitter.com/securitylabsxyz formerly https://twitter.com/dippixyz
Github: https://github.com/dippixyz

Contact:
doris@securitylabs.xyz - Doris Hernandez-Argueta
felipe@securitylabs.xyz - Felipe Flores
ata@securitylabs.xyz - Carlos Atahualpa Erazo

**Reply by momack2** [reply]
hey @Khriz01 - is SecurityLabs / Dippi building on libp2p, or adding new modules and implementations to libp2p?

**Reply by Khriz01** [reply]
Hey @momack2 - They're building a DePIN using libp2p to anonymously distribute sensitive data through their network.

**Comment by cewood** [comment]
**Project name:**
Amino DHT / go-libp2p-kad-dht

**Description:**
The Amino DHT is the canonical DHT solution in the libp2p space, and utilised throughout the libp2p ecosystem for tasks such as peer discovery, rendezvous, hole punching, and other things. Some example implementations that leverage it are kubo, go-ipfs, storm, edgevpn, go-libp2p, mintter, helia, Gear Node, nabu, parsec, HBX-CRYPTO-CHAT, js-libp2p, rust-libp2p-server, rust-ipfs, ceramic-one, Avail Node, edge-urid, bootstrap-node, ipns-server, and hyprspace.

**Github/Twitter link:**
https://github.com/libp2p/go-libp2p-kad-dht

**Contact:**
[@aschmahmann](https://github.com/aschmahmann)
[@cewood](https://github.com/cewood)
[@lidel](https://github.com/lidel)

**Comment by flcl42** [comment]
**Project name:**
.NET libp2p

**Description:**
Built from scratch with modern approaches and tech applied, the .NET libp2p unlocks p2p communication for apps written in C#. The implementation supports both TCP and QUIC, yamux as multiplexer and gossipsub 1.0 for pubsub. The library is in active development with current aim to achieve beta milestone. There are projects in progress that are already benefiting from it:
- It enables Ethereum execution client participation in Shutter network, Shutter fights bad MEV in Ethereum, the project is in testnet phase
- It makes several others apps in development able to join Ethereum Consensus Layer network

We are working on stability and API improvements, adding more tests and making libp2p interoperability tests all green. We want speed up adding other essential protocols and to try to add node to browser communication.
- Improve gossipsub 1.0 and finish gossipsub 1.1;
- Add browser build target and basic set of protocols needed;
- Add support for Kad DHT.

Additional funding will speed up development and will soon allow to build wider range of applications: IPFS based projects, more Ethereum apps, social media apps, exchanges, games. The goal is to have stable library with enough protocols and there is ongoing work of stabilization and fixing, but it’s important to have modern widely used protocols in the stack, this is currently postponed.

**Github/Twitter link:**

https://github.com/NethermindEth/dotnet-libp2p

**Contact:**
@flcl42
@rubo

**Comment by guillaumemichel** [comment]
Project name
---
Zikade

Description
---
Zikade is a new Go implementation of the [libp2p Kademlia DHT specification](https://github.com/libp2p/specs/tree/master/kad-dht), designed to be a successor to the existing implementation, [go-libp2p-kad-dht](https://github.com/libp2p/go-libp2p-kad-dht), which is now over [nine years old](https://github.com/libp2p/go-libp2p-kad-dht/commit/71d5f6fc8d16c458ae3d37b50f8477eff53e5390). 

The [go-libp2p-kad-dht](https://github.com/libp2p/go-libp2p-kad-dht) Kademlia implementation has reached a stage where adding significant features or improving algorithms has become increasingly difficult. It presents several core challenges:

 - Years of feature additions and core functionality iterations have increased performance but, as a side-effect, also added substantially to the complexity. It's now a daunting task for newcomers to understand the codebase and contribute meaningfully.
 - Due to the extensive parallelisation in different sections, writing stable unit tests has turned into a considerable challenge. Unreliable tests due to concurrency issues have become a bottleneck in the development pipeline.
 - A scarcity of reliable unit tests has made it difficult to run meaningful performance evaluations. Ambiguous results, like those seen with [Bitswap's Provider Search delay](https://github.com/ipfs/kubo/pull/9530), further cloud the picture.
 - The current architecture has accumulated a significant amount of technical debt. For instance, Kademlia should internally only handle Kademlia identifiers (256-bit bitstrings), but it currently uses strings, which is not only a deviation from the original design but also an unnecessary complexity.

In light of these challenges, Zikade aims to:

 - Make the code easier to understand and more straightforward to modify. 
 - Eliminate the flakiness associated with unconstrained concurrency in unit tests.
 - Provide a modular and extensible foundation for future DHT enhancements.
 - Allow the code to be easily tuned for performance optimisations.
 - Give more control over allocation and use of resources.
 - Prioritise execution of tasks such that 
   - local work is prioritized over incoming work
   - ongoing work is progressed before initating new work
   - work is bounded and backpressure is applied

You can read more about the motivation for revising the DHT implementation on the [IPFS Blog](https://blog.ipfs.tech/2023-09-amino-refactoring/).

According to some first [benchmarks](https://ipshipyard.slack.com/archives/C03L0G3B4RX/p1708678204714299) Zikade shows much better performances than [go-libp2p-kad-dht](https://github.com/libp2p/go-libp2p-kad-dht).

This work was performed by the [ProbeLab](https://probelab.io) team.

Github/Twitter link
---
https://github.com/probe-lab/zikade

Contact
---
@guillaumemichel 
@yiannisbot

**Comment by kaiserd** [comment]
**Project name:**
nim-libp2p

**Description:**
Nim-libp2p is a versatile libp2p implementation wirtten in [Nim](https://nim-lang.org/) used in  [Nimbus](https://nimbus.team/), [Waku](https://waku.org/), and [Codex](https://codex.storage/).

In Nimbus, an Ethereum client written in Nim, nim-libp2p plays a crucial role in facilitating peer discovery and communication within the Ethereum network. By leveraging nim-libp2p, Nimbus can efficiently connect to other Ethereum nodes, synchronize data,
and participate in consensus protocols, contributing to the robustness and decentralization of the Ethereum ecosystem.

Waku, a decentralized messaging protocol, utilizes nim-libp2p to create a resilient and censorship-resistant communication network.
With nim-libp2p, Waku nodes can discover and communicate with each other, ensuring privacy and security for message transmission.
This decentralized approach empowers users to exchange messages freely without relying on centralized servers.

Codex, a distributed storage platform, harnesses the capabilities of nim-libp2p to build a resilient and scalable storage network.

Overall, nim-libp2p serves as a foundational building block for decentralized applications and protocols.

**Github/Twitter link:**
https://github.com/vacp2p/nim-libp2p
https://twitter.com/vacp2p

**Contact:**
email: `ksr@status.im`
github: `kaiserd`
